### ISSUE TITLE
Custom operations in inductor

### ISSUE THREAD

**Issue by tbennun** [pytorch/pytorch#95021]
### 🚀 The feature, motivation and pitch

I'm working on a backend for Dynamo/Inductor and I would like to "catch" modules at a higher level than the lowered inductor graph. Lowering a high-level operator to ATen primitives misses out on semantics that could be used for efficient backend implementations. A few examples:
* `nn.Dropout` is immediately lowered to Philox random + `aten.where`, which makes certain assumptions
* `nn.BatchNorm` is reduced to its core operations, which disallows calling something like cuDNN
* Higher-level modules from PyTorch extensions cannot always be compiled to a single graph, due to their use of Pythonic features

I would still want to use the awesome optimizations and inlining features that inductor does on the original GraphModule, just not on every operator.

Ideally, the solution would entail taking components in their `nn.Module` level and replacing them with custom lowered operations that can be used internally. Dynamo will, in turn, [optionally] skip trying to trace through the module so as to not crash because of missing Pythonic features [if shape/type inference is provided].

### Alternatives

I tried wrapping modules with specific types with some wrapper class, which diverges in the `forward` function based on `dynamo.is_compiling()`. While verbose (and forcing me to redo shape inference), this allows me to change some internal behavior.

Following this step, none of the approaches I tried work. I tried several things, including registering custom operations via `torch.fx` and `torch.ops` (as in the custom ops tutorial https://pytorch.org/tutorials/advanced/torch_script_custom_ops.html). torch.fx wrapped functions are ignored, custom ops are lowered to ATen on the C++ level, and the lowered inductor graph always ends up with the primitives.

It is worth mentioning that in the ONNX exporter this used to be possible via registering custom schemata.

Thanks in advance!

**Comment by ezyang** [comment]
Re batch norm, you can always just turn off the decomp

**Comment by tbennun** [comment]
Thanks, how do I do that?

**Comment by ezyang** [comment]
I don't know how exactly you're integrating into Dynamo right now but there's a list of decomps in the registration API and you can change what is in that list.

**Comment by tbennun** [comment]
I made a dynamo backend as a function that:
* inspects and modifies the fx graph (turning some `call_module`s into `call_functions`)
* calls inductor.compile_fx with my own `inner_compile` so that I can get the lowered fwd/bwd graphs

Is it the intended use case? We won't support all high-level modules (that's the case for torch prims), so it makes sense to lower the graph. However, sometimes inductor lowers a bit too much.

**Comment by ngimel** [comment]
`compile_fx` calls aot_autograd with a standard set of decopositions https://github.com/pytorch/pytorch/blob/5783cee2a3a1457fc93b00a4a50e61ba02f148db/torch/_inductor/compile_fx.py#L458, you can change it (and perhaps we should expose it as an arg).

**Comment by tbennun** [comment]
@ngimel Thanks, now it's a PR (#95468). I'd also want the inverse (i.e., a do-not-decompose set of ops, or an opaque function). Even if I define a custom op in a separate library, the `call_function` in the original fx graph is still inlined/lowered/decomposed.

**Comment by eellison** [comment]
Will be fixed by https://github.com/pytorch/pytorch/pull/113823
